examples/cutting_stock_column_generation.jl

```diff
@@ -0,0 +1,194 @@
+and its dual solution is passed on to the pricing problem. The latter decides
+if there is a new pattern to include in the formulation or not; if so,
+it returns it to the master problem. The master is solved again, the new dual
+variables are given to the pricing problem, until there is no more pattern to
+generate from the pricing problem: all "interesting" patterns have been
+generated, and the master can take its optimal decision. In the implementation,
+the variables deciding how many times a pattern is chosen are called `θ`.
+
+For more information on column-generation techniques applied on the cutting
+stock problem, you can see:
+
+  * [Integer programming column generation strategies forthe cutting stock
+    problem and its variants](https://tel.archives-ouvertes.fr/tel-00011657/document)
+  * [Tackling the cutting stock problem](http://doi.org/10.5281/zenodo.3329388)
+"""
+function example_cutting_stock(; max_gen_cols::Int=5000)
+    maxwidth = 100.0
+    rollcost = 500.0
+    prices = Float64[167.0, 197.0, 281.0, 212.0, 225.0, 111.0, 93.0, 129.0, 108.0, 106.0, 55.0, 85.0, 66.0, 44.0, 47.0, 15.0, 24.0, 13.0, 16.0, 14.0]
+    widths = Float64[75.0, 75.0, 75.0, 75.0, 75.0, 53.8, 53.0, 51.0, 50.2, 32.2, 30.8, 29.8, 20.1, 16.2, 14.5, 11.0, 8.6, 8.2, 6.6, 5.1]
+    demand = Int[38, 44, 30, 41, 36, 33, 36, 41, 35, 37, 44, 49, 37, 36, 42, 33, 47, 35, 49, 42]
+    nwidths = length(prices)
+
+    n = length(widths)
+    ncols = length(widths)
+
+    # Initial set of patterns (stored in a sparse matrix: a pattern won't include many different cuts).
+    patterns = spzeros(UInt16, n, ncols)
+    for i in 1:n
+        patterns[i, i] = min(floor(Int, maxwidth / widths[i]), round(Int, demand[i]))
+    end
+
+    # Write the master problem with this "reduced" set of patterns.
+    # Not yet integer variables: otherwise, the dual values may make no sense
+    # (actually, GLPK will yell at you if you're trying to get duals for integer problems)
+    m = Model(with_optimizer(GLPK.Optimizer))
+    @variable(m, θ[1:ncols] >= 0)
+    @objective(m, Min,
+        sum(θ[p] * (rollcost - sum(patterns[j, p] * prices[j] for j in 1:n)) for p in 1:ncols)
+    )
+    @constraint(m, demand_satisfaction[j=1:n], sum(patterns[j, p] * θ[p] for p in 1:ncols) >= demand[j])
+
+    # First solve of the master problem.
+    optimize!(m)
+    if termination_status(m) != MOI.OPTIMAL
+        warn("Master not optimal ($ncols patterns so far)")
+    end
+
+    # Then, generate new patterns, based on the dual information.
+    while ncols - n <= max_gen_cols # Generate at most max_gen_cols columns.
+        if ! has_duals(m)
+            break
+        end
+
+        new_pattern = try
+            solve_pricing(dual.(demand_satisfaction), maxwidth, widths, rollcost, demand, prices)
+        catch
+            # At the final iteration, GLPK has dual values, but at least one of them is 0.0, and thus GLPK crashes.
```

Comment:
Why does a dual value of 0.0 cause GLPK to crash? That shouldn't happen.

Reply:
Actually, it's more GPLK.jl that gives an error (Cbc.jl too): 

```
ERROR: LoadError: AssertionError: dual >= 0.0
Stacktrace:
 [1] __assert_dual_sense__ at C:\Users\…\.julia\packages\LinQuadOptInterface\ZMx9f\src\solve.jl:210 [inlined]
 [2] get at C:\Users\…\.julia\packages\LinQuadOptInterface\ZMx9f\src\solve.jl:217 [inlined]
 [3] get(::MathOptInterface.Bridges.LazyBridgeOptimizer{GLPK.Optimizer,MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Bridges.AllBridgedConstraints{Float64}}}, ::MathOptInterface.ConstraintDual, ::MathOptInterface.ConstraintIndex{MathOptInterface.ScalarAffineFunction{Float64},MathOptInterface.GreaterThan{Float64}}) at C:\Users\…\.julia\packages\MathOptInterface\C3lip\src\Bridges\bridgeoptimizer.jl:236
 [4] get(::MathOptInterface.Utilities.CachingOptimizer{MathOptInterface.AbstractOptimizer,MathOptInterface.Utilities.UniversalFallback{JuMP._MOIModel{Float64}}}, ::MathOptInterface.ConstraintDual, ::MathOptInterface.ConstraintIndex{MathOptInterface.ScalarAffineFunction{Float64},MathOptInterface.GreaterThan{Float64}}) at C:\Users\…\.julia\packages\MathOptInterface\C3lip\src\Utilities\cachingoptimizer.jl:453
 [5] _moi_get_result(::MathOptInterface.Utilities.CachingOptimizer{MathOptInterface.AbstractOptimizer,MathOptInterface.Utilities.UniversalFallback{JuMP._MOIModel{Float64}}}, ::MathOptInterface.ConstraintDual, ::Vararg{Any,N} where N) at C:\Users\…\.julia\dev\JuMP\src\JuMP.jl:618
 [6] get(::Model, ::MathOptInterface.ConstraintDual, ::ConstraintRef{Model,MathOptInterface.ConstraintIndex{MathOptInterface.ScalarAffineFunction{Float64},MathOptInterface.GreaterThan{Float64}},ScalarShape}) at C:\Users\…\.julia\dev\JuMP\src\JuMP.jl:649
 [7] _constraint_dual(::ConstraintRef{Model,MathOptInterface.ConstraintIndex{MathOptInterface.ScalarAffineFunction{Float64},MathOptInterface.GreaterThan{Float64}},ScalarShape}) at C:\Users\…\.julia\dev\JuMP\src\constraints.jl:545
 [8] dual(::ConstraintRef{Model,MathOptInterface.ConstraintIndex{MathOptInterface.ScalarAffineFunction{Float64},MathOptInterface.GreaterThan{Float64}},ScalarShape}) at C:\Users\…\.julia\dev\JuMP\src\constraints.jl:538
 [9] _broadcast_getindex_evalf at .\broadcast.jl:625 [inlined]
 [10] _broadcast_getindex at .\broadcast.jl:598 [inlined]
 [11] getindex at .\broadcast.jl:558 [inlined]
 [12] copyto_nonleaf!(::Array{Float64,1}, ::Base.Broadcast.Broadcasted{Base.Broadcast.DefaultArrayStyle{1},Tuple{Base.OneTo{Int64}},typeof(dual),Tuple{Base.Broadcast.Extruded{Array{ConstraintRef{Model,C,Shape} where Shape<:AbstractShape where C,1},Tuple{Bool},Tuple{Int64}}}}, ::Base.OneTo{Int64}, ::Int64, ::Int64) at .\broadcast.jl:982
 [13] copy at .\broadcast.jl:836 [inlined]
 [14] materialize at .\broadcast.jl:798 [inlined]
 [15] #example_cutting_stock#3(::Int64, ::typeof(example_cutting_stock)) at C:\Users\…\Desktop\a.jl:126
 [16] example_cutting_stock() at C:\Users\…\Desktop\a.jl:88
 [17] top-level scope at C:\Users\…\Desktop\a.jl:175
```